##### Задача 1

Провести дисперсионный анализ для определения того, есть ли различия среднего роста среди взрослых футболистов, хоккеистов и штангистов. Даны значения роста в трех группах случайно выбранных спортсменов: 

football_players = [173, 175, 180, 178, 177, 185, 183, 182]  
hockey_players = [177, 179, 180, 188, 177, 172, 171, 184, 180]  
lifters = [172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170]  

In [1]:
import numpy as np

In [2]:
y1 = np.array([173, 175, 180, 178, 177, 185, 183, 182], dtype=np.float64)
y2 = np.array([177, 179, 180, 188, 177, 172, 171, 184, 180], dtype=np.float64)
y3 = np.array([172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170], dtype=np.float64)

n1 = y1.shape[0]
n2 = y2.shape[0]
n3 = y3.shape[0]

In [3]:
y1_mean = y1.mean()
y2_mean = y2.mean()
y3_mean = y3.mean()

y1_mean, y2_mean, y3_mean

(179.125, 178.66666666666666, 172.72727272727272)

In [4]:
y = np.concatenate([y1, y2, y3])
y_mean = y.mean()
y_mean

176.46428571428572

Найдём значения $S_F$ и $S_{res}$:

In [8]:
S2_F = n1 * (y1_mean - y_mean) ** 2 + n2 * (y2_mean - y_mean) ** 2 + n3 * (y3_mean - y_mean) ** 2
S2_F

253.9074675324678

In [11]:
S2_residual = ((y1 - y1_mean) ** 2).sum() + ((y2 - y2_mean) ** 2).sum() + ((y3 - y3_mean) ** 2).sum()
S2_residual

577.0568181818182

Найдем $S^2$ — сумму квадратов отклонений наблюдений от общего среднего:

In [13]:
S2 = np.sum((y - y_mean)**2)
S2

830.9642857142854

Проверим выполнение равенства $S^2 = S_F^2 + S_{res}^2:$

In [18]:
np.round(S2_F + S2_residual - S2, 10)

0.0

Найдем общую дисперсию:

In [27]:
k = 3
n = y.shape[0]
sigma2_general = S2 / (n - 1)
sigma2_general, n

(30.776455026455015, 28)

Найдем факторную дисперсию:

In [22]:
k1 = k - 1
sigma2_F = S2_F / k1
sigma2_F

126.9537337662339

Найдем остаточную дисперсию:

In [23]:
k2 = n - k
sigma2_residual = S2_residual / k2
sigma2_residual

23.08227272727273

Вычислим значение статистики $T$:

In [24]:
T = sigma2_F / sigma2_residual
T

5.500053450812598

Зафиксируем уровень значимости $\alpha = 0.05$ 
Для него и двух степеней свободы:  
$$df_{межд} = k - 1 = 3 - 1 = 2$$ и $$df_{внутр} = n - k = 28 - 3 = 25.$$  
найдём критическое значение $F_{crit}$ :

In [25]:
from scipy import stats

In [28]:
alpha = 0.05

F_crit = stats.f.ppf(1 - alpha, k1, k2)
F_crit

3.3851899614491687

Для данных значений $F_{крит} = 3.39$. Так как $F_H > F_{крит}$, различие среднего роста в трех группах статистически значимо.